AndroidHomeApplication/app/src/main/java/com/example/androidhomeapplication/data/repository/MoviesRepository.kt

```diff
@@ -0,0 +1,98 @@
+package com.example.androidhomeapplication.data.repository
+
+import com.example.androidhomeapplication.data.models.Genre
+import com.example.androidhomeapplication.data.models.Movie
+import com.example.androidhomeapplication.data.models.MovieDetails
+import com.example.androidhomeapplication.data.remote.RetrofitBuilder
+import com.example.androidhomeapplication.data.remote.response.*
+import com.example.androidhomeapplication.data.remote.services.ConfigurationService
+import com.example.androidhomeapplication.data.remote.services.MoviesService
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import retrofit2.Retrofit
+
+private const val SECURE_BASE_URL: String = "https://image.tmdb.org/t/p/"
+
+enum class ImageType(val size: String) {
+    POSTER(size = "w500"),
+    BACKDROP(size = "w780"),
+    PROFILE(size = "w185")
+}
+
+class MoviesRepository(
+    private val movieService: MoviesService,
+    private val configurationService: ConfigurationService
+) {
+
+
+    private val mutex = Mutex()
+    private var genres: Map<Long, Genre>? = null
+    private var configInfo: ConfigurationResponse? = null
+
+    suspend fun loadMovieById(movieId: Long): MovieDetails = coroutineScope {
+        val configurationInfo = async { getCachedConfigInfoOrLoad() }
+        val configurationInfoValue = configurationInfo.await()
+
+        val details = async { movieService.loadMovieDetails(movieId) }
+        val casts = async {
+            movieService.loadMovieCredits(movieId).cast.map { castResponse ->
+                castResponse.mapToActor(configurationInfoValue.getImageUrlByType(ImageType.PROFILE))
+            }
+        }
+
+        details.await().mapToMovieDetails(
+            configurationInfoValue.getImageUrlByType(ImageType.BACKDROP),
+            casts.await()
+        )
+    }
+
+    suspend fun loadMovies(queryString: String?, page: Int): List<Movie> = coroutineScope {
+        val movies = async { selectApiCallForMoviesData(queryString, page) }
+        val genresMap = async { getCachedGenresOrLoad() }
+        val configurationInfo = async { getCachedConfigInfoOrLoad() }
```

Comment:
here configurationInfo reference has no sense
but if u will put it before all other asyncs it will have sense